### PR TITLE
feat (billable-metrics): Add Prorated unique count service

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

This PR adds prorated unique count service

NOTE: There are some parts that are repeating in prorated unique count service and prorated sum service. Once we merge both services in feature branch, repeated parts will be handled in new PR

### CODE REVIEW

File: app/services/billable_metrics/advanced_aggregations/prorated_unique_count_service.rb [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+      end
+
+      def compute_pay_in_advance_aggregation
+        return BigDecimal(0) unless event
+        return BigDecimal(0) if event.properties.blank?
+
+        result_without_proration = super
+
+        number_of_days = to_datetime.in_time_zone(customer.applicable_timezone) -
+          event.timestamp.in_time_zone(customer.applicable_timezone)
```
Comment (jdenquin) [inline app/services/billable_metrics/advanced_aggregations/prorated_unique_count_service.rb:35]: Layout/MultilineOperationIndentation: Align the operands of an expression in an assignment spanning multiple lines.

File: app/services/billable_metrics/advanced_aggregations/prorated_unique_count_service.rb [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+      def handle_current_usage(result_without_proration, result_with_proration, is_pay_in_advance)
+        value_without_proration = result_without_proration.aggregation
+
+        if !is_pay_in_advance
+          result.aggregation = result_with_proration.negative? ? 0 : result_with_proration
+          result.current_usage_units = value_without_proration.negative? ? 0 : value_without_proration
+        elsif previous_event
+          result.current_usage_units = result_without_proration.current_usage_units
+          result.aggregation = result_without_proration.current_usage_units -
+            BigDecimal(previous_event.metadata['current_aggregation']) +
```
Comment (jdenquin) [inline app/services/billable_metrics/advanced_aggregations/prorated_unique_count_service.rb:168]: Layout/MultilineOperationIndentation: Align the operands of an expression in an assignment spanning multiple lines.

File: app/services/billable_metrics/advanced_aggregations/prorated_unique_count_service.rb [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            BigDecimal(previous_event.metadata['max_aggregation_with_proration'])
```
Comment (jdenquin) [inline app/services/billable_metrics/advanced_aggregations/prorated_unique_count_service.rb:169]: Layout/MultilineOperationIndentation: Align the operands of an expression in an assignment spanning multiple lines.

File: spec/services/billable_metrics/advanced_aggregations/prorated_unique_count_service_spec.rb [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+            removed_at:,
+            external_subscription_id: subscription.external_id,
+            billable_metric:,
+          )
+        end
+
+        before { new_quantified_event }
+
+        it 'returns the correct number' do
+          expect(result.aggregation).to eq((1 + (21.fdiv(31))).ceil(5))
```
Comment (jdenquin) [inline spec/services/billable_metrics/advanced_aggregations/prorated_unique_count_service_spec.rb:83]: Style/RedundantParentheses: Don't use parentheses around a method call.

File: spec/services/billable_metrics/advanced_aggregations/prorated_unique_count_service_spec.rb [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+          removed_at:,
+          external_subscription_id: subscription.external_id,
+          billable_metric:,
+        )
+      end
+
+      before { new_quantified_event }
+
+      it 'returns correct result' do
+        expect(result.aggregation).to eq((1 + (21.fdiv(31))).ceil(5))
```
Comment (jdenquin) [inline spec/services/billable_metrics/advanced_aggregations/prorated_unique_count_service_spec.rb:221]: Style/RedundantParentheses: Don't use parentheses around a method call.

**Review by vincent-pochet — COMMENTED** [review]
Just wondering if we could change the approach a bit, just to be able to share code between the pro-rated services:

What about something like:

```ruby
module BillableMetrics
  module ProratedAggregations
    class BaseService < BillableMetrics::Aggregations::BaseService
      def aggregation_without_proration
         @aggregation_without_proration ||= base_aggregator.aggregation
      end

      def pay_in_advance_aggregation
        return BigDecimal(0) unless event
        return BigDecimal(0) if event.properties.blank?

        number_of_days = to_datetime.in_time_zone(customer.applicable_timezone) -
                                       event.timestamp.in_time_zone(customer.applicable_timezone)

        proration_coefficient = number_of_days.fdiv(86_400).round.fdiv(period_duration)

        value = (result_without_proration * proration_coefficient).ceil(5)

        extend_event_metadata(value)

        value
      end

      # TODO: shared logic for all pro-rated aggregators

      private
      
      attr_reader :base_aggregator
    end
  end
end
```

```ruby

module BillableMetrics
  module ProratedAggregations
    class UniqueCountService < BillableMetrics::ProratedAggregations::BaseService
       def initialize(**args)
         @base_aggregator = BillableMetrics::Aggregations::UniqueCountService.new(**args)

         super(**args)
       end

      # TODO: the custom logic for the aggregation
    end
  end
end
```

**Pull request merged**